### ISSUE TITLE
`search_k` vs. number of index items

### ISSUE THREAD

**Issue by eddie-scio** [spotify/annoy#481]
Hi all, thanks for building a great library!  I understand at a high level the trade-off between accuracy and latency when using `search_k` (for a fixed memory constraint via `num_trees`), but a few questions:

1) If I set `search_k` to my index size, should I expect an exhaustive search?  I believe I've run into a case where I miss the true nearest neighbor when doing so.
2) Relatedly, if I increase `search_k`, is it expected that latency should continue to degrade, even past the full index size?  Does `k` itself factor in here?  I am also seeing this, e.g. for 100-tree index of 100k items of 128d, setting `search_k` to 500k and 1m gives a 2x increase in latency).

How can I better understand the "range" of `search_k`?

**Comment by erikbern** [comment]
On #1: no, that's not entirely right. Every tree contains every node, but the search algorithm searches all trees at the same time, so in theory an exhaustive search would be if you set search_k to n_trees * n_items. But if you truly want exhaustive search then you shouldn't use Annoy :)
On #2: yes, the search latency will be roughly linear in search_k, although it does some deduping of nodes when it computes distances, so it's possible once you go search_k > n_items then it might have some slight sublinear behavior

**Comment by eddie-scio** [comment]
Thanks for the quick and informative reply, Erik, good to know the tree search happens in parallel.

I was trying to quantify the degree of approximation -- is there an easy way to do so?  Any rules of thumb on inaccuracy as a function of these parameters?

**Comment by erikbern** [comment]
No, it's hard to really give any theoretical bounds – it's probably more empirical. Depends on the distribution of your data etc.

**Comment by erikbern** [comment]
> the tree search happens in parallel.

Actually, I didn't mean this happens in parallel as in multiple cores. What I meant was that it keeps a big priority queue of all trees, and goes through them serially in order using a single CPU.

**Comment by eddie-scio** [comment]
Hey Erik -- quick followup (didn't want to open up a whole new issue).  What impact (if any) does `k` have on the latency and accuracy, for everything else fixed?

**Comment by erikbern** [comment]
higher k means higher latency and higher accuracy!
